### ISSUE TITLE
TPC-H: run spark queries on databricks or EMR

### ISSUE THREAD

**Issue by mrocklin** [coiled/benchmarks#1189]
Maybe we're being dumb.  I think that running against a known-well-optimized system would help.

**Comment by milesgranger** [comment]
I've taken the liberty to run the PySpark queries on AWS EMR; using the same PySpark 3.4.1, number and size of instances. 
There is a gist here comparing them: https://gist.github.com/milesgranger/ef9a7e6ae0ff989bd5b41933bf02fe79

TL;DR
It appears we don't have much to worry about, pretty much on par with EMR. Some a bit faster, some a bit slower but all in the neighborhood.

Maybe a biased opinion, but it was way less ergonomic doing it with EMR. ~8-10min wait for clusters, and a boat load of toggles, drop downs and other stuff to set/choose.

| Query | EMR   | Coiled | Delta |
|-------|-------|--------|------------|
| 1     | 51.3  | 60.9   | +9.6       |
| 2     | 45.2  | 37.2   | -8.0       |
| 3     | 121.7 | 122.3  | +0.6       |
| 4     | 99.4  | 88.9   | -10.5      |
| 5     | 101.3 | 106.6  | +5.3       |
| 6     | 46.4  | 31.4   | -15.0      |
| 7     | 83.9  | 56.4   | -27.5      |

**Comment by phofl** [comment]
That’s good! The queries are a bit noisy  as well, so I agree that the
results look fine.

I would be interested in databricks as well, not sure what others think

Miles ***@***.***> schrieb am Do. 9. Nov. 2023 um 04:40:

**Comment by mrocklin** [comment]
Cool.  Thank you for doing this @milesgranger .  This helps me feel more confidence that we're not doing anything terrible.  I'm also curious about Databricks if that's easy (but maybe it isn't).

cc also @ntabris in case he's interested in the verification

**Comment by milesgranger** [comment]
Was able to run the same notebook/queries on DataBricks this morning. 
As a summary, it was way, way slower. So much in fact, I reached out to a previous colleague to look over my shoulder as he's presently using DataBricks a lot as his place of work. Nothing obvious stuck out to either of us why it was so slow compared to EMR and Coiled. Except that DataBricks heavily encourages the use of Delta Tables, and maybe that's why? Everything else was as normal, node types, cluster size, us-east-2 location, etc.

On top of that, I also trialed a couple different DataBricks specific configurations, using disk cache as indicated by the tip output, then with and without using [Photon](https://www.databricks.com/product/photon), all very slow and no discernible difference in performance. Maybe they just aren't very good with many small parquet files. 

Is it worth it trying to convert our data to a delta table then running again? Or maybe this workload of many parquet files is just part of the benchmark.

Updated table:

| Query | EMR   | Coiled | DataBricks |
|-------|-------|--------|------------|
| 1     | 51.3  | 60.9   | 359.9      |
| 2     | 45.2  | 37.2   | 696.9      |
| 3     | 121.7 | 122.3  | 1269.1     |
| 4     | 99.4  | 88.9   | 838.4      |
| 5     | 101.3 | 106.6  | 1550.4     |
| 6     | 46.4  | 31.4   | 656.7      |
| 7     | 83.9  | 56.4   | 1674.3     |

**Comment by mrocklin** [comment]
That seems strange.  We must be doing something wrong here, right?

cc'ing @vinooganesh in case he has suggestions on why we might be seeing much slower performance on Databricks than on EMR.

**Comment by milesgranger** [comment]
@vinooganesh, thanks! I should have been clearer maybe, but autoscaling was not enabled for any of these runs. On Coiled, EMR or DataBricks. All clusters were locked to 32 nodes (m6i.2xlarge).

The `get_spark_cluster`, is Coiled specific. The gist contains three notebooks, one for each vendor. The one ran on DataBricks is here: (second one down) https://gist.github.com/milesgranger/ef9a7e6ae0ff989bd5b41933bf02fe79#file-spark-databricks-tpch-testing-ipynb

**Comment by milesgranger** [comment]
Made a Loom going over how things look from my screen: https://www.loom.com/share/266ab81d9e3742eeaa05bf3e08bd3ec5?sid=a5da7fad-1530-4edc-b549-c914506e220d
(I said I'd run it again, but forgot. Anyhow did it offline and the run was similar, also before I click run in the video you can see it was ~370s on a prior run, also about the same as in the video.)

**Comment by mrocklin** [comment]
@ntabris if you have a moment.  In the beginning of Miles' video he points out various hardware configuration settings that are being done in Databricks, including disk.  If you can take a look and see if there are any differences in how Coiled is setting things up that might account for the difference that would be welcome.

**Comment by vinooganesh** [comment]
Hey @milesgranger - Thanks for uploading the loom. Commenting a few things here for record-keeping/sanity checking:
1. This cluster (Miles Granger's Cluster) is already started and running before you run the test
2. That cluster has 32 `m6i.2xlarge` workers and a `m6i.2xlarge` driver which you can see as started in the EC2 UI
3. Autoscaling is disabled, photo acceleration is enabled 
4. You have 1 on-demand driver, and 32 spot workers 
5. No custom sparkconf
6. I'm noticing on the Spark UI each of the executors have completed 36 tasks (though I can't tell if this is from just this job or you have run other jobs).
7. GC time doesn't look crazy

Can you confirm a few things?
1. Are you seeing any spot kills happening during your tests or are there 32 nodes up the entire time?
2.  How large is `s3://coiled-data/tpch/scale-1000/` and how many partitions? 
3. Is the any way I could access the Spark UI for your execution to see if something in your setup is causing this? I specifically want to see the query plan (You can go to the SQL tab in Spark UI and scroll to the bottom and expand the last item).

This does looks much, much slower than I'd expect (a 6x slowdown). Spark does struggle with a lot of files (a variant of the hadoop small files problem), so I want to make sure we're not hitting that.

**Comment by mrocklin** [comment]
> Spark does struggle with a lot of files (a variant of the hadoop small files problem), so I want to make sure we're not hitting that

To be clear, what's really odd to me is that Spark does ok when run on EMR and Coiled when using the same code on the same data.

**Comment by milesgranger** [comment]
> Are you seeing any spot kills happening during your tests or are there 32 nodes up the entire time?

No, and no other times either, first query is always pretty close to 6 minutes.

> How large is s3://coiled-data/tpch/scale-1000/ and how many partitions?

Total 10TB, each 'table' is many partuet files, ranging from ~20MiB to  ~128MiB, most closter to 128 .

> Is the any way I could access the Spark UI for your execution to see if something in your setup is causing this? I specifically want to see the query plan (You can go to the SQL tab in Spark UI and scroll to the bottom and expand the last item).

I'm happy to do this live with you. Kids are soon home, maybe in ~5hrs (8pm CET)

**Comment by phofl** [comment]
Just a FYI: ``scale-1000`` is 1TB not 10, so maybe a typo somewhere?

**Comment by vinooganesh** [comment]
Hi @milesgranger @phofl @mrocklin - I'll email you offline (and will include @brandon-kaplan too) to find some time

**Comment by mrocklin** [comment]
Question: I wonder if the `m` series instances are bad in some way for Databricks.  If it's easy we should maybe try the default instance type and then look at the cost difference.

**Comment by milesgranger** [comment]
> I wonder if the m series instances are bad in some way for Databricks. 

Good idea, I ran again with 32 i3.2xlarge nodes (twice the memory, same CPU as m6i.2xlarge), and after running the first query twice, it was even slower than on m6i.2xlarge instances: 11mins vs 6mins(!). 

Also tried upgrading to the latest software which used Spark 3.5.0 w/o any improvement (normal 6min run on query 1). I'm starting to question reality. Looking forward to our mob session on Monday, when it will no doubt run fast when people are present. :wink:
